### PR TITLE
Project 2 improve calculator

### PR DESCRIPTION
**Pull request by ggcarmi**

### CODE REVIEW

File: MyCalculator/CalculatorBrain.swift [hunk 1]
```diff
@@ -19,38 +19,41 @@ struct CalculatorBrain{
     private var accumulatorValue: Double?
     private var accumulatorDescription = " "
     
+    // this stack will represent the history
+    private var arrayOfLastElements = [ArrayStackMember]()
     
     // to store the first operand with the function
-    private var pendingBinaryOperation: PendingBinaryOperation?
+    var pendingBinaryOperation: PendingBinaryOperation?
```
Comment (oded-regev) [inline MyCalculator/CalculatorBrain.swift]: why did you change this from `private`?

File: MyCalculator/ViewController.swift [hunk 2]
```diff
@@ -58,26 +88,61 @@ class ViewController: UIViewController {
...
+        // M: current value of the display
+        variablesDictionary = ["M": displayValue]
+        
+        if let res = variablesDictionary!["M"] {
+            let isInteger = res.truncatingRemainder(dividingBy: 1) == 0
+            displayM.text = " M = " + (isInteger ? String(format: "%.0f", res) : String(res))
+        }
+        
+        userIsInMiddleOfTyping = false
+        displayValueToUpdate = brain.evaluate(using: variablesDictionary)
```
Comment (oded-regev) [inline MyCalculator/ViewController.swift:156]: I'm not sure we should call `evaluate()` here... please verify

File: MyCalculator/ViewController.swift [hunk 3]
```diff
@@ -58,26 +88,61 @@ class ViewController: UIViewController {
...
     override func viewDidLoad() {
         super.viewDidLoad()
         // Do any additional setup after loading the view, typically from a nib.
     }
 
+    // →M
+    
+    @IBAction func getM(_ sender: UIButton) {
+        // M: current value of the display
+        variablesDictionary = ["M": displayValue]
```
Comment (oded-regev) [inline MyCalculator/ViewController.swift]: Please note that you override the `var variablesDictionary` in this line. If in the future you'll have multiple vars this line will probably cause an unexpected bug since we would expect to **add** the new M entry to the dictionary and not override the entire dictionary

**Comment by oded-regev** [comment]
I just went over your code and also deeply tested the app. All in all it works great and as expected. The code looks good as well. I put some comments in the code, other than that there are 2 additional issues:

1. Please avoid using ! in the code, use `if let ...` or `guard let ..` 

2. The following scenarios are not working as expected (or at least I think it's not)

- 9, +, undo, 3, = (show 3 instead of 93)
- 3, + , 2, undo, undo, *, 2 (the operator * is ignored)
- 3, +, 2, =, undo, undo (still shows +)
- 2, ->M, 3 + M (displays 2 instead of M)
- 22.11 + 0.5 = , undo, 2, =, undo, undo (shows 0.5 which was deleted)

**Comment by ggcarmi** [comment]
1. i will check for the !, it will take some time to get use to it.
2. most of the cases is the same issue and i know what is the problem.
    i think it should act this way, because this is the way he build the calculator.
    in task 10 it says: if the user is in the middle of typing - delete the last character.
    if not - undo the last action.
    so when you going back and you get a number - you are not in the middle of typing.
    i can try to change it (using that boolean) - but im not sure how it will behave.
3.the example with the * is wired, i need to debug this.
4. your last example, it remember the 0.5 because you didnt erased it.
    your first undo cancel the = and not the 0.5

File: MyCalculator/ViewController.swift (hunk 3 above)
Comment (ggcarmi) [reply MyCalculator/ViewController.swift]: i know, this is the requirement in the task.
first, i just override the value of M,
but than i notice that the task says to drop the whole dictionary and use a new one.

Comment (oded-regev) [reply MyCalculator/ViewController.swift]: right.. so in `clear()` I would do something like:
`variablesDictionary = [String: Double]()`

instead of

`variablesDictionary = nil`

And in `getM` just do 
`variablesDictionary['M'] = displayValue`

**Comment by oded-regev** [comment]
About the "problematic scenarios", can you please let me know for each of the 5 scenarios I've mentioned, which one you think is OK and which one needed fix (and of course provide the fix when needed).

File: MyCalculator/ViewController.swift (hunk 3 above)
Comment (ggcarmi) [reply MyCalculator/ViewController.swift]: agree, i'll change it

File: MyCalculator/CalculatorBrain.swift (hunk 1 above)
Comment (ggcarmi) [reply MyCalculator/CalculatorBrain.swift]: i dont remember if i had a reason, i'll check it

File: MyCalculator/ViewController.swift (hunk 2 above)
Comment (ggcarmi) [reply MyCalculator/ViewController.swift:156]: according to the task it should be before, i'll check it

**Comment by ggcarmi** [comment]
9, +, undo, 3, = (show 3 instead of 93)
3, + , 2, undo, undo, *, 2 (the operator * is ignored)
3, +, 2, =, undo, undo (still shows +)
2, ->M, 3 + M (displays 2 instead of M)
22.11 + 0.5 = , undo, 2, =, undo, undo (shows 0.5 which was deleted)

1.  the problematic scenario
2. need to check - not sure
3. the problematic scenario - you need one more undo. the first one cancel the =,the second cancel 
   the 2, so its ok to stay with the +. the next undo will do it.
4. it display problem.the calculation are ok.
5. the problematic scenario - you need one more undo. same explanation as the above (3)
